cli/command/image/push.go

```diff
@@ -84,25 +112,84 @@ func RunPush(ctx context.Context, dockerCli command.Cli, opts pushOptions) error
 		All:           opts.all,
 		RegistryAuth:  encodedAuth,
 		PrivilegeFunc: requestPrivilege,
+		Platform:      platform,
 	}
 
 	responseBody, err := dockerCli.Client().ImagePush(ctx, reference.FamiliarString(ref), options)
 	if err != nil {
 		return err
 	}
 
+	defer func() {
+		for _, note := range notes {
+			fmt.Fprintln(dockerCli.Err(), "")
+			printNote(dockerCli, note)
+		}
+	}()
+
 	defer responseBody.Close()
 	if !opts.untrusted {
 		// TODO PushTrustedReference currently doesn't respect `--quiet`
 		return PushTrustedReference(dockerCli, repoInfo, ref, authConfig, responseBody)
 	}
 
 	if opts.quiet {
-		err = jsonmessage.DisplayJSONMessagesToStream(responseBody, streams.NewOut(io.Discard), nil)
+		err = jsonmessage.DisplayJSONMessagesToStream(responseBody, streams.NewOut(io.Discard), handleAux(dockerCli))
 		if err == nil {
 			fmt.Fprintln(dockerCli.Out(), ref.String())
 		}
 		return err
 	}
-	return jsonmessage.DisplayJSONMessagesToStream(responseBody, dockerCli.Out(), nil)
+	return jsonmessage.DisplayJSONMessagesToStream(responseBody, dockerCli.Out(), handleAux(dockerCli))
+}
+
+var notes []string
+
+func handleAux(dockerCli command.Cli) func(jm jsonmessage.JSONMessage) {
+	return func(jm jsonmessage.JSONMessage) {
+		b := []byte(*jm.Aux)
+
+		var stripped auxprogress.ManifestPushedInsteadOfIndex
+		err := json.Unmarshal(b, &stripped)
+		if err == nil && stripped.ManifestPushedInsteadOfIndex {
+			note := fmt.Sprintf("Not all multiplatform-content is present and only the available single-platform image was pushed\n%s -> %s",
+				red(stripped.OriginalIndex.Digest.String()),
+				green(stripped.SelectedManifest.Digest.String()))
+			notes = append(notes, note)
+		}
+
+		var missing auxprogress.ContentMissing
+		err = json.Unmarshal(b, &missing)
+		if err == nil && missing.ContentMissing {
+			note := `You're trying to push a manifest list/index which 
+				references multiple platform specific manifests, but not all of them are available locally
+				or available to the remote repository.
+
+				Make sure you have all the referenced content and try again.
+
+				You can also push only a single platform specific manifest directly by specifying the platform you want to push with the --platform flag.`
+			notes = append(notes, note)
+		}
+	}
+}
+
+func red(text string) string {
+	return fmt.Sprintf("\033[31m%s\033[0m", text)
+}
+
+func green(text string) string {
+	return fmt.Sprintf("\033[32m%s\033[0m", text)
+}
+
+func bold(text string) string {
+	return fmt.Sprintf("\033[1m%s\033[0m", text)
+}
+
+func printNote(dockerCli command.Cli, format string, args ...any) {
+	if _, isTty := term.GetFdInfo(dockerCli.Err()); isTty {
```

Comment:
Note that `dockerCli.Out()` has a `IsTerminal()` function that we can use. Not sure if that's fully correct though (is that correct if `STDERR` redirected, but `STDOUT` isn't?

We should consider providing this for _both_ `Out()` and `Err()` (I looked at that a few times, and wondered why we didn't)

Reply:
The most common case of this is when you redirect the stderr to `/dev/null`:

```
$ docker push dckr.woland.xyz/ububuu
Using default tag: latest
The push refers to repository [dckr.woland.xyz/ububuu]
aa21f24e1940: Layer already exists
latest: digest: sha256:17c24d16d63d2d089db74c2ed3e99c1ab0fd3f4f93c00b04afa8793fa793626c size: 424

[ NOTE ] Not all multiplatform-content is present and only the available single-platform image was pushed
sha256:e3f92abc0967a6c19d0dfa2d55838833e947b9d74edbcb0113e48535ad4be12a -> sha256:17c24d16d63d2d089db74c2ed3e99c1ab0fd3f4f93c00b04afa8793fa793626c

$ docker push dckr.woland.xyz/ububuu 2>/dev/null
Using default tag: latest
The push refers to repository [dckr.woland.xyz/ububuu]
aa21f24e1940: Layer already exists
latest: digest: sha256:17c24d16d63d2d089db74c2ed3e99c1ab0fd3f4f93c00b04afa8793fa793626c size: 424
```

Reply:
Yeah, I think there's other places where we do it wrong, and only look at `STDOUT`, but not at `STDERR` (but switch both based on `STDOUT`; here's a recent discussion that I still need to reply to, because I think the same is happening for `docker build` ; https://github.com/moby/moby/discussions/47755


Perhaps a good reason to look if we should make `Err()` provide thee same features as `Out()`, so that it's easier to check for each of the outputs if they have a TTY attached or not.

Reply:
In either case; not a blocker (should I mention that `isTty` should be `isTTY` ? 😂 🙈 )

Reply:
Will open a follow-up with the `stderr` change (it will have a bunch of updates in other places, so prefer to handle it in a separate PR).

Reply:
> `isTTY`

LOL, you're just merciless Sebastiaan 🤣 
...done!

Reply:
> so prefer to handle it in a separate PR

Yes; that's for sure something separate; and it's not a new thing, just that this is (I think) the first time we pay attention and check the right thing for a TTY 😂 

> OL, you're just merciless Sebastiaan 🤣


I'm the Human Linter! 😂